src/proper_erlang_abstract_code.erl

```diff
@@ -0,0 +1,3175 @@
+                true ->
+                    ?DEBUG("DISCARD!\n", []),
+                    false % discard; try again
+              %% ; false ->
+              %%       io:format("The linter found an error that should not "
+              %%                 "have occurred:\n ~p\n ~p\n", [Errors, T]),
+              %%       exit({bug, Errors, T})
+            end
+    end.
+
+call_linter(expr, Expr) ->
+    erl_lint:exprs([Expr], []);
+call_linter(forms, Forms) ->
+    erl_lint:module(Forms).
+
+all_hard_to_fix_errors(Errors) ->
+    lists:all(fun hard_error/1, Errors).
+
+%%% After some testing it seems harmless to treat all linter errors
+%%% as "hard". The "leakage" is negligible.
+
+%%% hard_error({_, erl_lint, illegal_bin_pattern}) -> true;
+%%% hard_error({_, erl_lint, {error, bittype_unit}}) -> true;
+%%% hard_error({_, erl_lint, illegal_map_key}) -> true;
+%%% hard_error({_, erl_lint, unsized_binary_in_bin_gen_pattern}) -> true;
+%%% Not yet seen:
+%%% illegal_guard_expr
+%%% utf_bittype_size_or_unit
+%%% {undefined_bittype, _}
+%%% unsized_binary_in_bin_gen_pattern
+%%% illegal_pattern_end
+%%% and probably more...
+%%% Not so hard errors, should have been taken care of:
+%%% ({_, erl_lint, {bittype_mismatch, _, _, _}})
+%%% ({_, erl_lint, unsized_binary_not_at_end})
+%%% hard_error({_, erl_lint, unsized_binary_in_bin_gen_pattern}) -> true;
+hard_error({_, erl_lint, _}) -> true.
+
+some_atoms() ->
+    ['', air, area, art, back, body, book, business, car, change,
+     child, city, community, company, country, day, door, education,
+     eye, face, fact, family, father, force, friend, game, girl,
+     government, group, guy, hand, head, health, history, home, hour,
+     house, idea, information, issue, job, kid, kind, law, level,
+     life, line, lot, man, member, minute, moment, money, month,
+     morning, mother, name, night, number, office, others, parent,
+     part, party, people, person, place, point, power, president,
+     problem, program, question, reason, research, result, right,
+     room, school, service, side, state, story, student, study,
+     system, teacher, team, thing, time, war, water, way, week, woman,
+     word, work, world, year].
+
+some_named_functions() ->
+    %% Do not include {'_', ...} since "fun _() -> _() end"
+    %% results in unbound variable.
+    [{'F1', 1}, {'F2', 0}, {'F3', 2}, {'F3', 3}].
+
+some_functions() ->
+    [{f1, 1}, {f1, 2}, {f2, 0}].
+
+some_records() ->
+    %% The first one is chosen as "guard record", which means that
+    %% all field initializations are guard expressions.
+    [{r1, [f1, f2]}, {r2, []}, {r3, [f1]}].
+
+%%% An arbitrary, small, collection of BIFs.
+auto_imported() ->
+    [{abs, 1}, {atom_to_list, 1}, {ceil, 1}, {erase, 1},
+     {exit, 1}, {group_leader, 2}, {is_function, 2},
+     %% Old BIFs:
+     {check_process_code, 2}, {get, 0}, {is_atom, 1}].
+
+some_types() ->
+    [{t1, 0}, {t2, 1}].
+
+predef_types() ->
```

Comment:
Two line breaks in this list are really unfortunate....
Similar in the list of the next function.